src/msm.rs

```diff
@@ -121,62 +351,73 @@ pub fn multiexp_serial<C: CurveAffine>(coeffs: &[C::Scalar], bases: &[C], acc: &
     }
 }
 
-/// Performs a small multi-exponentiation operation.
-/// Uses the double-and-add algorithm with doublings shared across points.
-pub fn small_multiexp<C: CurveAffine>(coeffs: &[C::Scalar], bases: &[C]) -> C::Curve {
-    let coeffs: Vec<_> = coeffs.iter().map(|a| a.to_repr()).collect();
-    let mut acc = C::Curve::identity();
-
-    // for byte idx
-    for byte_idx in (0..32).rev() {
-        // for bit idx
-        for bit_idx in (0..8).rev() {
-            acc = acc.double();
-            // for each coeff
-            for coeff_idx in 0..coeffs.len() {
-                let byte = coeffs[coeff_idx].as_ref()[byte_idx];
-                if ((byte >> bit_idx) & 1) != 0 {
-                    acc += bases[coeff_idx];
+pub fn best_multiexp<C: CurveAffine>(coeffs: &[C::Scalar], bases: &[C]) -> C::Curve {
+    // TODO: consider adjusting it with emprical data?
+    let batch_size = 64;
```

Comment:
This is likely to be too small for large batches

Assuming `c` with the number of buckets being 2^(c-1), I use the formula: `4c² - 16*c - 128`:
- https://github.com/mratsim/constantine/blob/58d8d2c/constantine/math/elliptic/ec_multi_scalar_mul_scheduler.nim#L260-L263
- For 2^10 = 1024 points, c = 9 (in Constantine, formula for c is gives result lower than logarithm to account for memory bandwidth limits): this is a batch size of 52
- For 2^26 = 67M points, c = 16, batch size of 640

Reply:
`4c² - 16*c - 128` makes this implementation much slower. I think it is because our parallelazing methods are different?

with k = 20

* `batch_size = 4c² - 16*c - 128` takes 618.459ms
* `batch_size = 64` takes 887.797ms

Reply:
Are the number swapped? because 618.459ms is 69.7% of 887.797ms hence the compute is faster.

Reply:
Yes copy pasted wrongly